cep-cache-output.md

```diff
@@ -0,0 +1,103 @@
+  build:
+    # note: in the current implementation, all other keys for build
+    #       are also valid, but unused
+    script: build.sh
+```
+
+<details> 
+  <summary>script build.sh default discussion</summary>
+We had some debate wether the cache output should _also_ default to `build.sh` or should not have any default value for the `script`. This is still undecided.
+</details>
+
+When computing variants and used variables, rattler-build looks at the union of a given output and the cache. That means, even if an output does not define any requirements, the cache would still add a variant for the `c_compiler`.
+
+When rattler-build executes the recipe, it will start by building the cache output that is appropriate for the current variant. This is computed by looking at all "used-variables" for the cache output and computing a "hash" for the cache. The build itself is executed in the same way as any other build.
+
+The variant keys that are injected at build time is the subset used by the cache output.
+
+When the cache build is done, the newly created files are moved outside of the `host-prefix`. Post-processing is not performed on the files beyond memoizing what files contain the `$PREFIX` (which is later replaced in binaries and text files with the actual build-prefix with a simple 1-to-1 replacement as the prefix-length are guaranteed to be equal).
+
+The cache _only_ restores files that were added to the prefix (conda-build also restored source files). Sources are re-created from scratch and no intermediate state is kept.
+
+<details> 
+  <summary>**Workaround for keeping intermediate build state**</summary>
+In order to keep state from the build around in the outputs, one could exercise the build in the $PREFIX and ignore the files in the outputs, or the build folder coule be placed in the `/tmp` folder:
+
+```bash
+cmake -B /tmp/mybuild -S .
+cmake --build /tmp/mybuild
+```
+
+Since this is not extremely convenient, we could discuss adding a "scratchpad" outside the `src` dir to rattler-build that will be kept between outputs and removed at the end of the whole build.
+</details>
+
+The new files can then be used in the outputs with the `build.files` key:
+
+```yaml
+outputs:
+  - package:
+      name: foo-headers
+
+    build:
+      files:
+        - include/**
+  - package:
+      name: libfoo
+    build:
+      files:
+        - lib/**
+
+  - package:
+      name: foo-devel
+    requirements:
+      run:
+        - ${{ pin_subpackage("libfoo") }}
+        - ${{ pin_subpackage("foo-headers") }}
+```
+
+Special care must be taken for `run-exports`. For example, the `${{ compiler('c') }}` package used to build the cache is going to have run-exports that need to be present at runtime for the package. To compute run-exports for the outputs, we use the union of the requirements - so virtually, the host & build dependencies of the cache are injected into the outputs and will attach their run exports to each output.
```

Comment:
I'm in favour of this, just noting that this diverges from conda-build, which currently does not inject REs from the global build-step to the outputs.

Reply:
I am not in favor of this because it is not an explicit behavior and not every output will need the run_exports from the cache. For example, if your devel package doesn't have any binaries, it doesn't need to depend on any compiler package exports. If you split your binaries into multiple outputs, each output may only depend on a part of what was used in `cache.requirements.host`.

I suspect that this design decision (implicit inheritance of run exports) will lead to more overdepending warnings (which are ignored by most maintainers). In contrast, if we retain the current behavior which is that each output must explicitly enumerate their dependencies, there will be overlinking errors (which are errors that stop the build) when dependencies are missing.